File: spike/route_circuit_registry.py
"""Bounded runtime storage for the frozen route-circuit v1 reducer."""

from dataclasses import dataclass
import threading
from typing import Optional

from route_circuit import (
    CircuitEvent,
    CircuitState,
    RouteCircuitKey,
    reduce_route_circuit,
)


@dataclass(frozen=True)
class RouteCircuitRegistryConfig:
    max_entries: int
    idle_ttl_ms: int


@dataclass(frozen=True)
class RouteCircuitRegistrySnapshot:
    key: RouteCircuitKey
    state: CircuitState
    last_touched_ms: int


def _validate_registry_config(registry_config, circuit_config):
    if registry_config.max_entries < 1:
        raise ValueError("max_entries must be positive")
    if registry_config.idle_ttl_ms < circuit_config.open_duration_ms:
        raise ValueError("idle_ttl_ms must cover open_duration_ms")


class RouteCircuitRegistry:
    """Thread-safe, deterministic TTL/LRU wrapper around route-circuit v1."""

    def __init__(self, circuit_config, registry_config):
        _validate_registry_config(registry_config, circuit_config)
        self._circuit_config = circuit_config
        self._registry_config = registry_config
        self._states = {}
        self._last_touched = {}
        self._last_event_ms: Optional[int] = None
        self._lock = threading.RLock()

    def _prune_idle(self, now_ms):
        expired = [
            key
            for key, touched_at in self._last_touched.items()
            if touched_at + self._registry_config.idle_ttl_ms <= now_ms
        ]
        for key in sorted(expired):
            self._states.pop(key, None)
            self._last_touched.pop(key, None)

    def _enforce_capacity(self, current_key):
        while len(self._states) > self._registry_config.max_entries:
            candidates = [key for key in self._states if key != current_key]
            if not candidates:
                candidates = list(self._states)
            evicted = min(
                candidates,
                key=lambda key: (self._last_touched[key], key),
            )
            self._states.pop(evicted, None)
            self._last_touched.pop(evicted, None)

    def apply(self, event):
        with self._lock:
            if event.now_ms < 0:
                raise ValueError("now_ms must not be negative")
            if self._last_event_ms is not None and event.now_ms < self._last_event_ms:
                raise ValueError("route-circuit registry time moved backwards")

            self._prune_idle(event.now_ms)
            self._states, decision = reduce_route_circuit(
                self._states,
                event,
                self._circuit_config,
            )
            if event.key in self._states:
                self._last_touched[event.key] = event.now_ms
            else:
                self._last_touched.pop(event.key, None)
            self._enforce_capacity(event.key)
            self._last_event_ms = event.now_ms
            return decision

    def clear(self):
        with self._lock:
            self._states.clear()
            self._last_touched.clear()
            self._last_event_ms = None

    def snapshot(self):
        with self._lock:
            return tuple(
                RouteCircuitRegistrySnapshot(
                    key=key,
                    state=self._states[key],
                    last_touched_ms=self._last_touched[key],
                )
                for key in sorted(self._states)
            )

    def __len__(self):
        with self._lock:
            return len(self._states)

File: app-tauri/src-tauri/src/route_circuit_registry.rs
//! Bounded runtime storage for the frozen route-circuit v1 reducer.

use crate::route_circuit::{
    reduce_route_circuit, CircuitConfig, CircuitDecision, CircuitEvent, CircuitState,
    CircuitStates, RouteCircuitKey,
};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteCircuitRegistryConfig {
    pub max_entries: usize,
    pub idle_ttl_ms: u64,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteCircuitRegistrySnapshot {
    pub key: RouteCircuitKey,
    pub state: CircuitState,
    pub last_touched_ms: u64,
}

#[derive(Clone, Debug)]
pub struct RouteCircuitRegistry {
    circuit_config: CircuitConfig,
    registry_config: RouteCircuitRegistryConfig,
    states: CircuitStates,
    last_touched: BTreeMap<RouteCircuitKey, u64>,
    last_event_ms: Option<u64>,
}

impl RouteCircuitRegistry {
    pub fn new(
        circuit_config: CircuitConfig,
        registry_config: RouteCircuitRegistryConfig,
    ) -> Result<Self, String> {
        if registry_config.max_entries < 1 {
            return Err("max_entries must be positive".into());
        }
        if registry_config.idle_ttl_ms < circuit_config.open_duration_ms {
            return Err("idle_ttl_ms must cover open_duration_ms".into());
        }
        Ok(Self {
            circuit_config,
            registry_config,
            states: CircuitStates::new(),
            last_touched: BTreeMap::new(),
            last_event_ms: None,
        })
    }

    fn prune_idle(&mut self, now_ms: u64) {
        let expired: Vec<RouteCircuitKey> = self
            .last_touched
            .iter()
            .filter(|(_, touched_at)| {
                touched_at.saturating_add(self.registry_config.idle_ttl_ms) <= now_ms
            })
            .map(|(key, _)| key.clone())
            .collect();
        for key in expired {
            self.states.remove(&key);
            self.last_touched.remove(&key);
        }
    }

    fn enforce_capacity(&mut self, current_key: &RouteCircuitKey) {
        while self.states.len() > self.registry_config.max_entries {
            let mut candidates: Vec<&RouteCircuitKey> = self
                .states
                .keys()
                .filter(|key| *key != current_key)
                .collect();
            if candidates.is_empty() {
                candidates.extend(self.states.keys());
            }
            let evicted = candidates
                .into_iter()
                .min_by_key(|key| (self.last_touched.get(*key).copied().unwrap_or(0), *key))
                .expect("over-capacity registry must contain an entry")
                .clone();
            self.states.remove(&evicted);
            self.last_touched.remove(&evicted);
        }
    }

    pub fn apply(&mut self, event: &CircuitEvent) -> Result<CircuitDecision, String> {
        if self
            .last_event_ms
            .is_some_and(|last_event_ms| event.now_ms < last_event_ms)
        {
            return Err("route-circuit registry time moved backwards".into());
        }

        self.prune_idle(event.now_ms);
        let (states, decision) = reduce_route_circuit(&self.states, event, &self.circuit_config)?;
        self.states = states;
        if self.states.contains_key(&event.key) {
            self.last_touched.insert(event.key.clone(), event.now_ms);
        } else {
            self.last_touched.remove(&event.key);
        }
        self.enforce_capacity(&event.key);
        self.last_event_ms = Some(event.now_ms);
        Ok(decision)
    }

    pub fn clear(&mut self) {
        self.states.clear();
        self.last_touched.clear();
        self.last_event_ms = None;
    }

    pub fn snapshot(&self) -> Vec<RouteCircuitRegistrySnapshot> {
        self.states
            .iter()
            .map(|(key, state)| RouteCircuitRegistrySnapshot {
                key: key.clone(),
                state: state.clone(),
                last_touched_ms: self.last_touched[key],
            })
            .collect()
    }

    pub fn len(&self) -> usize {
        self.states.len()
    }

    pub fn is_empty(&self) -> bool {
        self.states.is_empty()
    }
}
